### PR TITLE
Enable logging of tunnel messages

### PR DESCRIPTION
**Pull request by roman-dvorak**
Hi, I would like to enable [tunnel](https://mavlink.io/en/services/tunnel.html) mavlink messages in onboard logger. 

This PR is solved by creating a uORB tunnel message, translating a MAVLINK message into a uORB and logging it.

### CODE REVIEW

**Comment by dagar** [comment]
I'm just curious, what's the use case here?

Is it actually useful dumping these payloads into ulg? Alternatively you could do things like write the tunneled data straight to a file on the SD card.

**Comment by roman-dvorak** [comment]
Hi @dagar, yes, it is our use-case. To logging data from payload. 

> Alternatively you could do things like write the tunneled data straight to a file on the SD card.
You mean data directly from mavlink? Is there some "standard way" for this, or I should create custom mavlink logger?

Thank you!

**Comment by dagar** [comment]
> You mean data directly from mavlink? Is there some "standard way" for this, or I should create custom mavlink logger?

Yes having the mavlink receiver directly log the reconstructed data directly to a file. I'm not even sure it's necessarily a good idea, unless the tunneled data payloads make sense to reconstruct as a single file that's useful later. I'm imagining something like a ublox log file you could then open in u-center directly rather than having to pull it out of the ulg file and reconstruct.

I don't have a preference, whatever works.

File: src/modules/logger/logged_topics.cpp [hunk 1]
```diff
@@ -84,6 +84,7 @@ void LoggedTopics::add_default_topics()
 	add_topic("tecs_status", 200);
 	add_topic("test_motor", 500);
 	add_topic("trajectory_setpoint", 200);
+	add_topic("tunnel", 10);
```
Comment (dagar) [inline src/modules/logger/logged_topics.cpp:87]: Can we skip this in the default set, maybe add it as an option?

File: src/modules/logger/logged_topics.cpp [hunk 2]
```diff
@@ -172,6 +173,7 @@ void LoggedTopics::add_high_rate_topics()
 	add_topic("manual_control_setpoint");
 	add_topic("rate_ctrl_status", 20);
 	add_topic("sensor_combined");
+	add_topic("tunnel");
```
Comment (dagar) [inline src/modules/logger/logged_topics.cpp:176]: How about a new option just for tunnel? High-rate is really for analyzing vehicle flight and controller performance, we shouldn't overload it with other messages unless it's useful in that context.

**Comment by roman-dvorak** [comment]
Well, I'm open to solving this problem differently, somehow more conceptually. If it were possible to involve the opinions of more people to this problem. 

A huge advantage of logging this data in ulog is, that it can be very easily compared with other data from ulog file. Parameters such as position (location, orientation in space, airspeed, vibrations, RTK position, also drag of motor)

Next advantage is, that there is the same time base as through all the logged values. 

We need some universial interface for connecting some payolad and logging data from them. At the same time, we need to transfer data to the ground from payload (.and prehaps back). 

The data format is not always same. It depends on the type of payload. It can be one simple value, some matrix, or ... I dont know. 

The data will be decoded based on some "decoder" - it can be some python script. 

> I'm imagining something like a ublox log file you could then open in u-center directly rather than having to pull it out of the ulg file and reconstruct.

This path is very similar to logging GPS data. GPS raw dump data are also logged to ulog and then decoded to ubx format. 

The assumed payload has UART output with mavlink packets. However, it can also be a PX4 application producing mavlink data.

**Comment by dagar** [comment]
Screen Shot 2020-12-16 at 7 26 55 PM

To resolve this build failure you need to add the new .msg to `msg/tools/uorb_rtps_message_ids.yaml`.

File: msg/CMakeLists.txt [hunk 3]
```diff
@@ -133,6 +133,7 @@ set(msg_files
 	trajectory_waypoint.msg
 	transponder_report.msg
 	tune_control.msg
+	tunnel.msg
```
Comment (dagar) [inline msg/CMakeLists.txt:136]: Can we give this a more descriptive name? Maybe something like telemetry_tunnel or mavlink_tunnel?

File: msg/tunnel.msg [hunk 4]
```diff
@@ -0,0 +1,6 @@
+uint64 timestamp	# time since system start (microseconds)
+uint16 payload_type  # A code that identifies the content of the payload (0 for unknown, which is the default). If this code is less than 32768, it is a 'registered' payload type and the corresponding code should be added to the MAV_TUNNEL_PAYLOAD_TYPE enum. Software creators can register blocks of types as needed. Codes greater than 32767 are considered local experiments and should not be checked in to any widely distributed codebase.*/
```
Comment (dagar) [inline msg/tunnel.msg:2]: Add the enums as constants in here. We want to technically keep Mavlink completely separate from uORB. 

https://mavlink.io/en/messages/common.html#MAV_TUNNEL_PAYLOAD_TYPE

File: msg/tunnel.msg [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+uint8 target_system  # System ID (can be 0 for broadcast, but this is discouraged)*/
```
Comment (dagar) [inline msg/tunnel.msg:3]:
```suggestion
uint8 target_system  # System ID (can be 0 for broadcast, but this is discouraged)
```

File: msg/tunnel.msg [hunk 6]
```diff
@@ -0,0 +1,6 @@
...
+uint8 target_component  # Component ID (can be 0 for broadcast, but this is discouraged)*/
+uint8 payload_length  #  Length of the data transported in payload*/
```
Comment (dagar) [inline msg/tunnel.msg:5]:
```suggestion
uint8 payload_length  #  Length of the data transported in payload
```

**Comment by dagar** [comment]
> Well, I'm open to solving this problem differently, somehow more conceptually. If it were possible to involve the opinions of more people to this problem.

I think in this case it would be logging it directly to the "real" filetype based on payload type (https://mavlink.io/en/messages/common.html#MAV_TUNNEL_PAYLOAD_TYPE). Anyway, I'm fine with ulg if it's useful.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 7]
```diff
@@ -1857,6 +1862,24 @@ MavlinkReceiver::handle_message_obstacle_distance(mavlink_message_t *msg)
...
+	mavlink_msg_tunnel_decode(msg, &mavlink_tunnel);
+
+	tunnel_s tunnel{};
+
+	tunnel.timestamp = hrt_absolute_time();
+	tunnel.payload_type = mavlink_tunnel.payload_type;
+	tunnel.target_system = mavlink_tunnel.target_system;
+	tunnel.target_component = mavlink_tunnel.target_component;
+	tunnel.payload_length = mavlink_tunnel.payload_length;
+	memcpy(tunnel.payload, mavlink_tunnel.payload, sizeof(tunnel.payload));
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp:1878]: Probably just being paranoid, but you could add a static assert that the payloads are the same size. The mavlink message is technically work in progress and can change.

Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp:1878]:
```suggestion
	memcpy(tunnel.payload, mavlink_tunnel.payload, sizeof(tunnel.payload));
	static_assert(sizeof(tunnel.payload) == sizeof(mavlink_tunnel.payload), "tunnel.payload size mismatch");
```

File: msg/tools/uorb_rtps_message_ids.yaml [hunk 8]
```diff
@@ -387,4 +387,6 @@ rtps:
   - msg: estimator_odometry
     id: 180
     alias: vehicle_odometry
+  - msg: 	mavlink_tunnel
```
Comment (roman-dvorak) [inline msg/tools/uorb_rtps_message_ids.yaml:390]: Is it okay to add it to the end and not alphabetically?

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 7 above)
Comment (roman-dvorak) [reply src/modules/mavlink/mavlink_receiver.cpp:1878]: Yes, this is a better way. Thanks.

**Comment by roman-dvorak** [comment]
so, @dagar, I will summary the edits related to this PR:

- tunnel.msg renamed tu mavlink_tunnel.msg
- msg deffinition added to `msg/tools/uorb_rtps_message_ids.yaml `
- Added payload type enum to message deffinition
- logging is only in default logger. Logging is possible to enable trought parametr 'MAV_DUMP_TUNNEL'

**Comment by roman-dvorak** [comment]
@dagar, please, can you help me to resolve failing checks?

**Comment by dagar** [comment]
> @dagar, please, can you help me to resolve failing checks?

Builds restarted, it doesn't look like your changes could have caused those failures. Can you make sure this is up to date with current master?

**Comment by roman-dvorak** [comment]
I have merged it with upstream. Then I tested it with a CUAV V5+ autopilot and our test mavlink sender. The log is attached. 

[mavlink_tunnel_logs.zip](https://github.com/PX4/PX4-Autopilot/files/5860373/mavlink_tunnel_logs.zip)

File: src/modules/logger/logged_topics.cpp [hunk 9]
```diff
@@ -177,6 +177,13 @@ void LoggedTopics::add_default_topics()
 	if (gps_dump_comm == 1) {
 		add_topic("gps_dump");
 	}
+
+	int32_t mav_dump_tunnel = 0;
+	param_get(param_find("MAV_DUMP_TUNNEL"), &mav_dump_tunnel);
```
Comment (dagar) [inline src/modules/logger/logged_topics.cpp:182]: The analogous way to handle this is to have mavlink only republish the tunnel messages if `MAV_DUMP_TUNNEL` is enabled. Then on the logger side checking the param value before adding the topic is a small optimization, but not strictly required. In general try to keep modules contained to their own parameter space.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 10]
```diff
@@ -238,6 +238,11 @@ MavlinkReceiver::handle_message(mavlink_message_t *msg)
 		handle_message_obstacle_distance(msg);
 		break;
 
+	case MAVLINK_MSG_ID_TUNNEL:
+		handle_message_mavlink_tunnel(msg);
```
Comment (dagar) [inline src/modules/mavlink/mavlink_receiver.cpp:242]: Nitpick, please keep the naming convention consistent (handle_message_NAME).

handle_message_mavlink_tunnel -> handle_message_tunnel

File: src/modules/mavlink/params.c [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+ * This is useful for logging communication of payload and GCS
+ *
+ * @reboot_required true
+ * @boolean
+ * @group MAVLink
+ *
+ * @value 0 Disable
+ * @value 1 Enable
+ */
+PARAM_DEFINE_INT32(MAV_DUMP_TUNNEL, 0);
```
Comment (dagar) [inline src/modules/mavlink/params.c:46]: There's already a file with mavlink parameters. Can you move this? 

https://github.com/PX4/PX4-Autopilot/blob/master/src/modules/mavlink/mavlink_params.c

**Comment by mrpollo** [comment]
@roman-dvorak can you please address the comments by @dagar above and also rebase master on top of your branch once again? thanks for your contribution.

**Comment by roman-dvorak** [comment]
@mrpollo Yes, sure.. sorry for my delay. I'll do it tomorrow.

File: src/modules/logger/logged_topics.cpp (hunk 9 above)
Comment (roman-dvorak) [reply src/modules/logger/logged_topics.cpp:188]: Yes, that makes sense. The message is now translated according to the parameter.

File: src/modules/mavlink/params.c (hunk 11 above)
Comment (roman-dvorak) [reply src/modules/mavlink/params.c:46]: Moved.

**Comment by roman-dvorak** [comment]
I didn't have time for that before... sorry, 

@dagar I think, that it is ready. Thanks.

File: src/modules/mavlink/mavlink_receiver.cpp [hunk 12]
```diff
@@ -240,6 +240,19 @@ MavlinkReceiver::handle_message(mavlink_message_t *msg)
 		handle_message_obstacle_distance(msg);
 		break;
 
+	case MAVLINK_MSG_ID_TUNNEL: {
+			// Re-publish MAVLINK message to uORB only if I want to record it
+			// (parameter MAV_DUMP_TUNNEL is enabled)
+			int32_t mav_dump_tunnel = 0;
+			param_get(param_find("MAV_DUMP_TUNNEL"), &mav_dump_tunnel);
```
Comment (BazookaJoe1900) [inline src/modules/mavlink/mavlink_receiver.cpp:247]: `param_find()` is heavy function, its does long for loop.
using it on periodical function is not recommended. you should use `DEFINE_PARAMETERS `section

Comment (BazookaJoe1900) [reply src/modules/mavlink/mavlink_receiver.cpp:247]: anyway, why not to enable/disable the logging by new bit on 'SDLOG_PROFILE'? that seems more sutiable

Comment (roman-dvorak) [reply src/modules/mavlink/mavlink_receiver.cpp]: @BazookaJoe1900 Thanks for this information. I used section `DEFINE_PARAMETERS`

I did it in the style of enabling of the GPS dump. Should I fix it there? If you want, I'll do it in the next PR after merging this.

Comment (BazookaJoe1900) [reply src/modules/mavlink/mavlink_receiver.cpp]: My opinion is that it should be added as bit on SDLOG_PROFILE, and not on MAV_DUMP_TUNNEL.
its better to do it under this pr, and not in another one.

**Comment by BazookaJoe1900** [comment]
@dagar  is it possible to transfer & log variable length urob message?
so data won't be always 128 bytes?

**Comment by dagar** [comment]
> @dagar is it possible to transfer & log variable length urob message?
> so data won't be always 128 bytes?

I've considered that idea in the past (eg sensor_accel_fifo), but so far it's been at odds with the goal of uORB ultimately being type safe and memory safe end to end. I'm certainly not opposed to the idea if someone can find a safe and simple way to do it.

I also try to keep an eye on general performance across the typical platforms. In most of these cases the cost of publishing a slightly larger uORB message is actually fairly minimal unless it's happening at a high rate (kHz).

Example, the time it takes to copy vehicle_local_position vs sensor_gyro
 - vehicle_local_position (165 bytes), ~1.9 us/copy
 - sensor_gyro (40 bytes), ~ 1.5 us /copy
Screenshot from 2021-02-07 10-29-04

File: src/modules/mavlink/mavlink_receiver.cpp (hunk 12 above)
Comment (dagar) [reply src/modules/mavlink/mavlink_receiver.cpp]: I concur, mavlink republishing `mavlink_tunnel` at all times is cheap enough to not worry about it. Please extend `SDLOG_PROFILE` instead and we'll finally get this merged.

**Comment by roman-dvorak** [comment]
I'll do it by the end of the week

**Comment by roman-dvorak** [comment]
Hi, 
on request of @BazookaJoe1900 I redesigned it to enable logging with the `SDLOG_PROFILE` parameter bit (11th bit). Can you review it, please?

**Comment by kaklik** [comment]
> > Well, I'm open to solving this problem differently, somehow more conceptually. If it were possible to involve the opinions of more people to this problem.
> 
> I think in this case it would be logging it directly to the "real" filetype based on payload type (https://mavlink.io/en/messages/common.html#MAV_TUNNEL_PAYLOAD_TYPE). Anyway, I'm fine with ulg if it's useful.

I think that handling the payload data as a real filetype based on TUNNEL_PAYLOAD_TYPE results in very complex PX4 code. Because each PAYLOD_TYPE should have its own data decoder, which needs to be updated to payload firmware. 
The main use case of this feature is the logging of measurements performed by the payload.  The values could be a variety of atmospheric quantities (solid particles, radiation, gases, contamination, temperature, humidity, etc.)

**Comment by kaklik** [comment]
> > @dagar is it possible to transfer & log variable length urob message?
> > so data won't be always 128 bytes?
> 
> I've considered that idea in the past (eg sensor_accel_fifo), but so far it's been at odds with the goal of uORB ultimately being type safe and memory safe end to end. I'm certainly not opposed to the idea if someone can find a safe and simple way to do it.
> 
> I also try to keep an eye on general performance across the typical platforms. In most of these cases the cost of publishing a slightly larger uORB message is actually fairly minimal unless it's happening at a high rate (kHz).
> 
> Example, the time it takes to copy vehicle_local_position vs sensor_gyro
> 
>     * vehicle_local_position (165 bytes), ~1.9 us/copy
> 
>     * sensor_gyro (40 bytes), ~ 1.5 us /copy
>       Screenshot from 2021-02-07 10-29-04

In the intended use cases, I think it is not really probable to log the 128 bytes of tunnel data much more frequently than the ones of Hertz. It is caused by relatively slow time response of chemical or radiation sensors.

**Comment by kaklik** [comment]
Hi, is there some active issue that blocks the merging of this PR to upstream, please?

**Review by dagar — APPROVED** [review]

**Pull request merged**